### PR TITLE
[WIP] [CR] Unsafe npc shots through player to hit zombie

### PR DESCRIPTION
**Pull request by Wyattwashere**
#### Summary

SUMMARY: Bugfixes "Fixed the issue of NPCs shooting through you to get to zombies."

#### Purpose of change

Fixes issue #29358 which often resulted in the injury or death of a player due to NPCs ignoring targeting preferences

#### Describe the solution

Added a line in in npcmove to place the player in the NPC's ally cache, where before the line was missing.

#### Describe alternatives you've considered

Changing how the NPCs track allies could be doable, probably.

#### Additional context

I added a test file for ranged combat, and am trying to get a better idea of how to do the actual test and check to see if the targets are being acquired how I think they should be.

### CODE REVIEW

**Comment by I-am-Erk** [comment]
So wait. How many characters have died to a single missing line of code, do we think?

File: src/npcmove.cpp [hunk 1]
```diff
@@ -1572,6 +1573,7 @@ bool npc::wont_hit_friend( const tripoint &tar, const item &it, bool throwing )
             return false;
         }
     }
+
```
Comment (AMurkin) [inline src/npcmove.cpp]:
```suggestion
```

File: tests/npc_targetting_test.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+{
+    standard_npc dude("TestCharacter", {}, 3, 8, 8, 8, 8);
+    dude.weapon = item("1911");
+    return dude;
+}
+
+TEST_CASE("NPC Shoots enemy without player in the way.", "[.ranged]")
+{
+    monster zed(mtype_id("mon_zombie"));
+
```
Comment (AMurkin) [inline tests/npc_targetting_test.cpp]:
```suggestion
```

File: src/npcmove.cpp [hunk 3]
```diff
@@ -432,6 +432,7 @@ void npc::assess_danger()
         } else if( is_friendly( g->u ) ) {
             float min_danger = assessment >= NPC_DANGER_VERY_LOW ? NPC_DANGER_VERY_LOW : -10.0f;
             assessment = std::max( min_danger, assessment - player_diff * 0.5f );
+            ai_cache.friends.emplace_back(g->shared_from(g->u));
```
Comment (AMurkin) [inline src/npcmove.cpp]:
```suggestion
            ai_cache.friends.emplace_back( g->shared_from( g->u ) );
```

**Comment by ZhilkinSerg** [comment]
Would you apply suggestions from code review please?

**Comment by ZhilkinSerg** [comment]
I've removed test, so it would compile (we can add one later).

**Comment by Mooses2k** [comment]
Does this code affect vehicle turrets? One of the only things that can kill my heavy power armor guy is his own laser cannon turret.

**Comment by ZhilkinSerg** [comment]
Yes, it should affect turrets which are shooting using fake friendly npc.

**Pull request merged**